Fixed #12090 -- Show admin actions on the edit pages too

mgaligniana opened the pull request:
Hi!

- It fixes [ticket 12090](https://code.djangoproject.com/ticket/12090)
- **Discussion thread**: https://forum.djangoproject.com/t/admin-actions-also-in-the-object-details-view/22927/4
- To run new tests please do: `./runtests.py admin_views.test_actions.AdminDetailActionsTest`
- When `save_on_top` is `True` looks like:

![image](https://github.com/django/django/assets/16822952/dffaf334-ce46-4774-83f7-7d929018fb37)

mgaligniana commented on django/contrib/admin/options.py:
```diff
@@ -973,10 +975,10 @@ def action_checkbox(self, obj):
         return getattr(func, "short_description", capfirst(name.replace("_", " ")))
 
-    def _get_base_actions(self):
+    def _get_base_actions(self, klass_actions):
```
To avoid duplicate code, I re-write this method to accept the `actions` or `change_actions` by param. Probably there is a better way to implement this.

mgaligniana commented on django/contrib/admin/options.py:
```diff
@@ -973,10 +975,10 @@ def action_checkbox(self, obj):
         return getattr(func, "short_description", capfirst(name.replace("_", " ")))
 
-    def _get_base_actions(self):
+    def _get_base_actions(self, class_actions):
```
To avoid duplicate code, I re-write this method to accept the `actions` or `change_actions` by param. Probably there is a better way to implement this.

djm commented on django/contrib/admin/templates/admin/includes/edit_actions.html:
```diff
@@ -0,0 +1,7 @@
+    <input type="submit" name="{{action.1}}" value="{{action.2}}">
+    {% endfor %}
+</div>
+{% endif %}
```
First off: love this PR, it's very much a desire-able feature for us (@yunojuno) and we have been close to implementing it for ourselves a few times now. We keep admin inline actions for the kind of day-to-day support tasks that only developers should carry out and running them currently is a bit of a hassle.

I see the UI here though the problematic bit, and probably the reason this will meet resistance. On certain models we have say 10 inline actions - and with the UI as it is that would print 10 buttons to the screen. We'd be inundated with buttons.

A potential alternative is to go with the same UI as the listing pages. i.e A `select` based dropdown with a single button to carry out the action.

mgaligniana replied:
> A potential alternative is to go with the same UI as the listing pages

Finally applied!

djm commented:
Another point that is worth mentioning: Django Admin Inline Actions have the `description` attribute which can be used to [give a more friendly human-readable name to the action](https://docs.djangoproject.com/en/4.1/ref/contrib/admin/actions/#writing-action-functions).

e.g 
```
@admin.action(description='Mark selected stories as published')
def make_published(modeladmin, request, queryset):
    queryset.update(status='p')
```

The problem this poses is that this attribute was always designed to be used as part of an action against multiple instances of a given model and thus over the years people (me..) would have used names written in the plural-form like "Regenerate contracts", or "Refresh indexes". 

Therefore using those names in the context of the edit page would cause them to not make sense because it's a singular object being "actioned" rather than multiple. Others would have followed this naming pattern too, especially as the documentation example uses the wording "Mark selected stories as published" which would not work in singular form.

In an ideal world, inline actions would need two variables to do a good job of this - as Django does in many other places (`verbose_name` vs `verbose_name_plural`) etc. But this poses another problem in that the known-plural version is currently called `description` so it's not as simple as adding a new `description_plural` parameter.

felixxm commented:
@mgaligniana Do you have time to keep working on this?

mgaligniana commented:
@felixxm Yes! This week I'm going to add some progress! 🙏 Sorry, I've been busy these months.

mgaligniana commented on django/contrib/admin/options.py:
```diff
@@ -1556,11 +1598,13 @@ def response_post_save_change(self, request, obj):
         return self._response_post_save(request, obj)
 
-    def response_action(self, request, queryset):
+    def response_action(self, request, queryset, change):
```
Now `queryset` param could be an object. Should we rename it? Maybe "selected"? 🤔

mgaligniana commented:
> Another point that is worth mentioning: Django Admin Inline Actions have the `description` attribute which can be used to [give a more friendly human-readable name to the action](https://docs.djangoproject.com/en/4.1/ref/contrib/admin/actions/#writing-action-functions).
> 
> e.g
> 
> ```python
> @admin.action(description='Mark selected stories as published')
> def make_published(modeladmin, request, queryset):
>     queryset.update(status='p')
> ```
> 
> The problem this poses is that this attribute was always designed to be used as part of an action against multiple instances of a given model and thus over the years people (me..) would have used names written in the plural-form like "Regenerate contracts", or "Refresh indexes".
> 
> Therefore using those names in the context of the edit page would cause them to not make sense because it's a singular object being "actioned" rather than multiple. Others would have followed this naming pattern too, especially as the documentation example uses the wording "Mark selected stories as published" which would not work in singular form.
> 
> In an ideal world, inline actions would need two variables to do a good job of this - as Django does in many other places (`verbose_name` vs `verbose_name_plural`) etc. But this poses another problem in that the known-plural version is currently called `description` so it's not as simple as adding a new `description_plural` parameter.

Has been passed almost 1 year! Sorry djm, but finally I've returned to work on this! 🙏 

As I defined other different variable for detail view: `detail_actions`. The `description` could be written in singular way! That solve the problem?

mgaligniana commented on django/contrib/admin/options.py:
```diff
@@ -1038,11 +1054,25 @@ def get_actions(self, request):
+
+    def get_detail_actions(self, request):
+        """
+        Same as `get_actions` but for `detail_actions`.
```
Not sure if it is a good docstring. The only difference with `get_actions` is that it uses `_get_detail_actions`, but we need another method in case user wants overwrite it

mgaligniana commented on django/contrib/admin/options.py:
```diff
@@ -1051,10 +1081,22 @@ def get_action_choices(self, request, default_choices=models.BLANK_CHOICE_DASH):
             choices.append(choice)
         return choices
 
+    def get_detail_action_choices(
```
Same here

mgaligniana commented on django/contrib/admin/options.py:
```diff
@@ -1051,10 +1081,22 @@ def get_action_choices(self, request, default_choices=models.BLANK_CHOICE_DASH):
+        self, request, default_choices=models.BLANK_CHOICE_DASH
+    ):
+        """
+        Same as `get_action_choices` but for `detail_actions`.
```
Same here

nessita reviewed (commented):
Added some initial thoughts.

nessita commented on django/contrib/admin/options.py:
```diff
@@ -672,6 +672,9 @@ class ModelAdmin(BaseModelAdmin):
     actions_selection_counter = True
     checks_class = ModelAdminChecks
 
+    detail_actions = ()
```
@mgaligniana My suggestion would be to start a new topic in the forum, in the Django Internals category, referencing the old mailing list thread, I believe the forum is a bit more active than the mailing list. Thanks!

mgaligniana replied:
Done! https://forum.djangoproject.com/t/admin-actions-also-in-the-object-details-view/22927

Please let me know if I'm missing something or any sentence/grammar fix!

Thank you!

mgaligniana commented on tests/admin_views/admin.py:
```diff
@@ -409,7 +409,13 @@ def redirect_to(modeladmin, request, selected):
-    buf = StringIO("This is the content of the file")
+    from django.db.models.query import QuerySet
+
+    if isinstance(selected, QuerySet):
```
🎗️ This handle should be documented? Or should the action get `modeladmin, request, selected, object` as params? With `selected` as `None` when is change view or `object` as None when is list view 🤔

mgaligniana commented on django/contrib/admin/options.py:
```diff
@@ -1019,6 +1022,19 @@ def _get_base_actions(self):
+        actions = []
+        base_actions = (self.get_action(action) for action in self.detail_actions or [])
+        # get_action might have returned None, so filter any of those out.
+        base_actions = [action for action in base_actions if action]
```
Marked as resolved, no longer applies after changes discussed in django forum*

mgaligniana commented:
Hi @nessita!

I applied the `description_plural` paramenter and added a new test.

Two questions/doubts:
- I pass down the `cache` parameter (to know if its a change or list view) until `_get_action_description` but not sure if it is the best approach. What do you think? Is there another clean way?
- Following the forum discussion: The next step would be remove the `Delete` button?  (as it was replaced by the action inside the dropdown)

Thank you! And sorry the delay!

mgaligniana commented:
Hi @nessita! Just as a 🎗️, no rush. Any thoughts about https://github.com/django/django/pull/16012#issuecomment-1741925611 and https://github.com/django/django/pull/16012#discussion_r1342654481 ? To move forward and add more tests cases if its necessary. Thank you!

nessita commented:
> Hi @nessita! Just as a 🎗️, no rush. Any thoughts about [#16012 (comment)](https://github.com/django/django/pull/16012#issuecomment-1741925611) and [#16012 (comment)](https://github.com/django/django/pull/16012#discussion_r1342654481) ? To move forward and add more tests cases if its necessary. Thank you!

Hi! This this definitely in my ToDo list, but I've been prioritizing release blockers for 5.0 and some DjangoCon PRs (to capitalize on the momentum). I should be able to review this soon! Thank you for your patience :100:

sarahboyce commented on django/contrib/admin/templates/admin/change_form.html:
```diff
@@ -34,7 +40,8 @@
 <form {% if has_file_field %}enctype="multipart/form-data" {% endif %}{% if form_url %}action="{{ form_url }}" {% endif %}method="post" id="{{ opts.model_name }}_form" novalidate>{% csrf_token %}{% block form_top %}{% endblock %}
-<div>
+<div id="changeform">
+{% if action_form %}{% admin_actions %}{% endif %}
```
Have you considered including a hidden input here, something like:
```html
<input type="hidden" name="{{ action_checkbox_name }}" value="{{ obj.pk|unlocalize }}">
```

The idea being that
```python
selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
```
returns a list of one item being the object PK for the change form.

Then you can still handle everything as previously rather than now having to handle the single option case? I'm not sure if this will work but I think it's worth a try? What do you think?

mgaligniana replied:
> ```
> action_checkbox_name
> ```

Thank you Sarah! I've tried that option but I couldn't make it work, and also I had to send via the context the `ACTION_CHECKBOX_NAME` and the `obj`

I finally solved the issue adding this line: https://github.com/django/django/pull/16012/files#diff-e77dfa6935de3fac94356ffd1029a679f8aa23f2979fe9b31aea7f93e1dccad3R1887

With that `selected` will be always a `queryset` in case of the detail view: a QuerySet of 1 object. IMO easier to handle when you define your custom action. And no backwards compatibility breaks

Let me know what do you think!

sarahboyce replied:
I had a bit of a further look into it and sorry there was some mistakes in the suggestion but I can get this to work with the following
```diff
diff --git a/django/contrib/admin/actions.py b/django/contrib/admin/actions.py
index 805ae0feca..314fe1becb 100644
--- a/django/contrib/admin/actions.py
+++ b/django/contrib/admin/actions.py
@@ -31,11 +31,6 @@ def delete_selected(modeladmin, request, queryset):
     opts = modeladmin.model._meta
     app_label = opts.app_label
 
-    # Handle the case when delete action is called from change view and
-    # it is a single object
-    if not isinstance(queryset, QuerySet):
-        queryset = modeladmin.model.objects.filter(id=queryset.id)
-
     # Populate deletable_objects, a data structure of all related objects that
     # will also be deleted.
     (
diff --git a/django/contrib/admin/options.py b/django/contrib/admin/options.py
index 9961ddd729..ceb18a6609 100644
--- a/django/contrib/admin/options.py
+++ b/django/contrib/admin/options.py
@@ -1615,7 +1615,7 @@ class ModelAdmin(BaseModelAdmin):
         """
         return self._response_post_save(request, obj)
 
-    def response_action(self, request, queryset, change):
+    def response_action(self, request, queryset, change=False):
         """
         Handle an admin action. This is called if a request is POSTed to the
         changelist or changeform; it returns an HttpResponse if the action was
@@ -1652,26 +1652,25 @@ class ModelAdmin(BaseModelAdmin):
         # If the form's valid we can handle the action.
         if action_form.is_valid():
             action = action_form.cleaned_data["action"]
+            select_across = action_form.cleaned_data["select_across"]
             func = self.get_actions(request)[action][0]
-            if not change:
-                select_across = action_form.cleaned_data["select_across"]
-                # Get the list of selected PKs. If nothing's selected, we can't
-                # perform an action on it, so bail. Except we want to perform
-                # the action explicitly on all objects.
-                selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
-                if not selected and not select_across:
-                    # Reminder that something needs to be selected or nothing will
-                    # happen
-                    msg = _(
-                        "Items must be selected in order to perform "
-                        "actions on them. No items have been changed."
-                    )
-                    self.message_user(request, msg, messages.WARNING)
-                    return None
 
-                if not select_across:
-                    # Perform the action only on the selected objects
-                    queryset = queryset.filter(pk__in=selected)
+            # Get the list of selected PKs. If nothing's selected, we can't
+            # perform an action on it, so bail. Except we want to perform
+            # the action explicitly on all objects.
+            selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
+            if not selected and not select_across:
+                # Reminder that something needs to be selected or nothing will happen
+                msg = _(
+                    "Items must be selected in order to perform "
+                    "actions on them. No items have been changed."
+                )
+                self.message_user(request, msg, messages.WARNING)
+                return None
+
+            if not select_across:
+                # Perform the action only on the selected objects
+                queryset = queryset.filter(pk__in=selected)
 
             response = func(self, request, queryset)
 
(main-3.12) sarahboyce@sarahboyce-ThinkPad-E15-Gen-4:~/Documents/django/tests$ git diff
diff --git a/django/contrib/admin/actions.py b/django/contrib/admin/actions.py
index 805ae0feca..314fe1becb 100644
--- a/django/contrib/admin/actions.py
+++ b/django/contrib/admin/actions.py
@@ -31,11 +31,6 @@ def delete_selected(modeladmin, request, queryset):
     opts = modeladmin.model._meta
     app_label = opts.app_label
 
-    # Handle the case when delete action is called from change view and
-    # it is a single object
-    if not isinstance(queryset, QuerySet):
-        queryset = modeladmin.model.objects.filter(id=queryset.id)
-
     # Populate deletable_objects, a data structure of all related objects that
     # will also be deleted.
     (
diff --git a/django/contrib/admin/options.py b/django/contrib/admin/options.py
index 9961ddd729..ceb18a6609 100644
--- a/django/contrib/admin/options.py
+++ b/django/contrib/admin/options.py
@@ -1615,7 +1615,7 @@ class ModelAdmin(BaseModelAdmin):
         """
         return self._response_post_save(request, obj)
 
-    def response_action(self, request, queryset, change):
+    def response_action(self, request, queryset, change=False):
         """
         Handle an admin action. This is called if a request is POSTed to the
         changelist or changeform; it returns an HttpResponse if the action was
@@ -1652,26 +1652,25 @@ class ModelAdmin(BaseModelAdmin):
         # If the form's valid we can handle the action.
         if action_form.is_valid():
             action = action_form.cleaned_data["action"]
+            select_across = action_form.cleaned_data["select_across"]
             func = self.get_actions(request)[action][0]
-            if not change:
-                select_across = action_form.cleaned_data["select_across"]
-                # Get the list of selected PKs. If nothing's selected, we can't
-                # perform an action on it, so bail. Except we want to perform
-                # the action explicitly on all objects.
-                selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
-                if not selected and not select_across:
-                    # Reminder that something needs to be selected or nothing will
-                    # happen
-                    msg = _(
-                        "Items must be selected in order to perform "
-                        "actions on them. No items have been changed."
-                    )
-                    self.message_user(request, msg, messages.WARNING)
-                    return None
 
-                if not select_across:
-                    # Perform the action only on the selected objects
-                    queryset = queryset.filter(pk__in=selected)
+            # Get the list of selected PKs. If nothing's selected, we can't
+            # perform an action on it, so bail. Except we want to perform
+            # the action explicitly on all objects.
+            selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
+            if not selected and not select_across:
+                # Reminder that something needs to be selected or nothing will happen
+                msg = _(
+                    "Items must be selected in order to perform "
+                    "actions on them. No items have been changed."
+                )
+                self.message_user(request, msg, messages.WARNING)
+                return None
+
+            if not select_across:
+                # Perform the action only on the selected objects
+                queryset = queryset.filter(pk__in=selected)
 
             response = func(self, request, queryset)
 
@@ -1881,17 +1880,14 @@ class ModelAdmin(BaseModelAdmin):
         actions = self.get_actions(request)
         if request.method == "POST":
             if actions and request.POST.get("action", ""):
-                action_failed = False
                 response = self.response_action(
                     request,
-                    self.get_queryset(request).filter(id=object_id),
-                    change=not add,
+                    self.get_queryset(request),
+                    change=True
                 )
                 if response:
                     return response
                 else:
-                    action_failed = True
-                if action_failed:
                     # Redirect back to the changelist page to avoid resubmitting the
                     # form if the user refreshes the browser or uses the "No, take
                     # me back" button on the action confirmation page.
@@ -1968,7 +1964,7 @@ class ModelAdmin(BaseModelAdmin):
 
         # Build the action form and populate it with available actions.
         if actions and not add:
-            action_form = self.action_form(auto_id=None)
+            action_form = self.action_form(initial={"change": True}, auto_id=None)
             action_form.fields["action"].choices = self.get_action_choices(
                 request, change=True
             )
@@ -1987,6 +1983,7 @@ class ModelAdmin(BaseModelAdmin):
             "to_field": to_field,
             "media": media,
             "action_form": action_form,
+            "action_checkbox_name": helpers.ACTION_CHECKBOX_NAME,
             "inline_admin_formsets": inline_formsets,
             "errors": helpers.AdminErrorList(form, formsets),
             "preserved_filters": self.get_preserved_filters(request),
@@ -2085,7 +2082,7 @@ class ModelAdmin(BaseModelAdmin):
         ):
             if selected:
                 response = self.response_action(
-                    request, queryset=cl.get_queryset(request), change=False
+                    request, queryset=cl.get_queryset(request)
                 )
                 if response:
                     return response
@@ -2109,7 +2106,7 @@ class ModelAdmin(BaseModelAdmin):
         ):
             if selected:
                 response = self.response_action(
-                    request, queryset=cl.get_queryset(request), change=False
+                    request, queryset=cl.get_queryset(request)
                 )
                 if response:
                     return response
diff --git a/django/contrib/admin/templates/admin/change_form.html b/django/contrib/admin/templates/admin/change_form.html
index 43dc9fef3c..c07a166fe5 100644
--- a/django/contrib/admin/templates/admin/change_form.html
+++ b/django/contrib/admin/templates/admin/change_form.html
@@ -1,5 +1,5 @@
 {% extends "admin/base_site.html" %}
-{% load i18n admin_urls static admin_modify %}
+{% load i18n l10n admin_urls static admin_modify %}
 
 {% block extrahead %}{{ block.super }}
 <script src="{% url 'admin:jsi18n' %}"></script>
@@ -41,7 +41,7 @@
 {% endblock %}
 <form {% if has_file_field %}enctype="multipart/form-data" {% endif %}{% if form_url %}action="{{ form_url }}" {% endif %}method="post" id="{{ opts.model_name }}_form" novalidate>{% csrf_token %}{% block form_top %}{% endblock %}
 <div id="changeform">
-{% if action_form %}{% admin_actions %}{% endif %}
+{% if action_form %}{% admin_actions %}<form method="post">{% csrf_token %}<input type="hidden" name="{{ action_checkbox_name }}" value="{{ object_id|unlocalize }}"></form>{% endif %}
 {% if is_popup %}<input type="hidden" name="{{ is_popup_var }}" value="1">{% endif %}
 {% if to_field %}<input type="hidden" name="{{ to_field_var }}" value="{{ to_field }}">{% endif %}
 {% if save_on_top %}{% block submit_buttons_top %}{% submit_row %}{% endblock %}{% endif %}
diff --git a/tests/admin_views/test_actions.py b/tests/admin_views/test_actions.py
index 77a67b0e6e..c8c5f8c51c 100644
--- a/tests/admin_views/test_actions.py
+++ b/tests/admin_views/test_actions.py
@@ -645,6 +645,7 @@ class AdminDetailActionsTest(TestCase):
                 "email": "foo@bar.com",
                 "_save": "Save",
                 "action": "",
+                ACTION_CHECKBOX_NAME: [self.s1.pk],
             },
         )
         external_subscriber = ExternalSubscriber.objects.get()
@@ -663,6 +664,7 @@ class AdminDetailActionsTest(TestCase):
                 "email": "foo@bar.com",
                 "_save": "Save",
                 "action": "external_mail",
+                ACTION_CHECKBOX_NAME: [self.s1.pk],
             },
         )
         self.assertEqual(len(mail.outbox), 1)
@@ -679,7 +681,7 @@ class AdminDetailActionsTest(TestCase):
 
         response = self.client.post(
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
-            {"action": "download"},
+            {"action": "download", ACTION_CHECKBOX_NAME: [self.s1.pk]},
         )
         content = b"".join(list(response))
         self.assertEqual(
@@ -696,7 +698,7 @@ class AdminDetailActionsTest(TestCase):
         self.user.user_permissions.add(permission)
         response = self.client.post(
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
-            {"action": "delete_selected"},
+            {"action": "delete_selected", ACTION_CHECKBOX_NAME: [self.s1.pk]},
         )
         self.assertEqual(response.status_code, 200)
         self.assertContains(
@@ -708,6 +710,7 @@ class AdminDetailActionsTest(TestCase):
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
             {
                 "action": "delete_selected",
+                ACTION_CHECKBOX_NAME: [self.s1.pk],
                 "post": "yes",
             },
         )
@@ -718,7 +721,7 @@ class AdminDetailActionsTest(TestCase):
         # User doesn't have the permission to run the custom action.
         response = self.client.post(
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
-            {"action": "custom_action"},
+            {"action": "custom_action", ACTION_CHECKBOX_NAME: [self.s1.pk]},
         )
         self.assertEqual(response.status_code, 302)
 
@@ -732,7 +735,7 @@ class AdminDetailActionsTest(TestCase):
         self.user.user_permissions.add(permission)
         response = self.client.post(
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
-            {"action": "custom_action"},
+            {"action": "custom_action", ACTION_CHECKBOX_NAME: [self.s1.pk]},
         )
         self.assertEqual(response.status_code, 200)
         self.assertEqual(response.content, b"OK")
diff --git a/tests/admin_views/tests.py b/tests/admin_views/tests.py
index e0a4926b91..82917bd4dc 100644
--- a/tests/admin_views/tests.py
+++ b/tests/admin_views/tests.py
@@ -4841,6 +4841,7 @@ class AdminInheritedInlinesTest(TestCase):
         )
         names = name_re.findall(response.content)
         names.remove(b"csrfmiddlewaretoken")
+        names.remove(b"csrfmiddlewaretoken")
         # make sure we have no duplicate HTML names
         self.assertEqual(len(names), len(set(names)))
```

sarahboyce commented on tests/admin_views/test_actions.py:
```diff
@@ -529,3 +529,155 @@ def test_model_admin_no_delete_permission_externalsubscriber(self):
+        )
+        self.assertEqual(response.status_code, 200)
+
+    def test_delete_action_in_detail_view(self):
```
Can we add a test around permissions? Either in here or in `AdminActionsPermissionTests` 🤔

sarahboyce reviewed (commented):
Hello @mgaligniana 👋 thank you so much for this and being patient with us. 
To give you a summary of what I think needs to be done to move this forward:

- we need a release note and some doc updates, perhaps a section in here (https://docs.djangoproject.com/en/5.0/ref/contrib/admin/actions/). 
- I tried to give an option that might resolve some of your comments (see https://github.com/django/django/pull/16012#discussion_r1551148654), this might not work but I would be interested in whether you have investigated it and your thoughts.

mgaligniana commented:
@sarahboyce Thank you Sarah!! Sorry for not answering previosly! For sure, I'm going back to work on this and update the Tarcker when is ready! 💪

mgaligniana commented:
This is how looks like the documentation I added:

![image](https://github.com/user-attachments/assets/3b8e7a67-8c34-456b-a1d4-e1e9cf31ac60)

mgaligniana commented:
I haven't added SeleniumTest but I added the label to check I didn't break anything. Let me know if add some would be required!

mgaligniana commented:
I installed my Django branch version. I recorded a happy path video:

[happy_path_actions_in_detail_view.webm](https://github.com/user-attachments/assets/4b817857-494e-41dd-a349-e3145ca90013)

I think we have things to test and define, but we are close! :crossed_fingers:

sarahboyce commented on django/contrib/admin/options.py:
```diff
@@ -1072,7 +1077,7 @@ def _filter_actions_by_permissions(self, request, actions):
         return filtered_actions
 
-    def get_actions(self, request):
+    def get_actions(self, request, change=False):
```
I was trying to think if we can figure out if this is coming from the change view by looking at the request (to try and reduce some of the change kwarg additions
I don't have something on the top of my head, but still thinking about it

mgaligniana replied:
🎗️ I got it using the `resolver_match`

sarahboyce commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -149,6 +149,24 @@ That code will give us an admin change list that looks something like this:
+.. note::
+
+    In this case ``queryset`` will be a :class:`~django.db.models.query.QuerySet` of one object.
+
```
We need to add a `.. versionchanged:: 5.2` note at the bottom here

sarahboyce reviewed (commented):
Thank you for the updates @mgaligniana :star: 
I've been testing this today and have a few suggestions

We also need a release note in Django 5.2 :+1:

smiling-watermelon commented on django/contrib/admin/actions.py:
```diff
@@ -29,6 +31,11 @@ def delete_selected(modeladmin, request, queryset):
 
+    # Handle the case when delete action is called from change view and
+    # it is a single object
+    if not isinstance(queryset, QuerySet):
+        queryset = modeladmin.model.objects.filter(id=queryset.id)
```
Not an expert in how filtering works in admin, but shouldn't this be a `pk` filter?
Not all models have `id` field, but all models (well, all non-abstract models) should have a `pk` field.

Just tried it out locally by making an instance of my Admin page, and I get:
```
In [10]: admin.model.objects.filter(id=uuid4())
Out [10]:
# Long stack trace
FieldError: Cannot resolve keyword 'id' into field. Choices are: ...
```

And changing this filter to `pk` results in:
```
In [11]: admin.model.objects.filter(pk=uuid4())
Out[11]: <QuerySet []>
```
_Note: Model used for this example is using UUIDField for Primary Key_

So perhaps this should be:
```suggestion
        queryset = modeladmin.model.objects.filter(pk=queryset.pk)
```
Where `queryset` is actually an instance of type `modeladmin.model`.
And in that case, maybe it's better to change the "typecheck" the other way around, and make it:
```suggestion
# in imports:
from django.db.models import Model

# In `delete_selected` method:
    if isinstance(queryset, Model):
        queryset = modeladmin.model.objects.filter(pk=queryset.pk)
```

(I know Django is not really a typed project, and all, but in this case typecheckers will be certain that `queryset` object has the `pk` attribute, which is a plus IMO)

mgaligniana replied:
Good finding! I'll try continue this PR as soon as possible and I'll add a test for this case. Thank you!

mgaligniana replied:
So after applying the Sarah's suggest I think this was solved!

mgaligniana commented:
Just an update: This week I have some days free so I'll have time to continue the work here! 🤞 💪

mgaligniana commented:
I've applied all the comments! + added couple of tests (I updated the PR description with more details)

Let me know any test or change to add! 💪 I will not disappear again and I will try to finish it!

knyghty commented:
@mgaligniana the label is just so the selenium tests run. With that in mind I've also added the screenshots label.

I'm not sure if a selenium test is needed or not. I guess it would be nice to write a test that selects an action and runs it, though I'm not sure if it's just duplicating the tests that are already written.

I will try to do an accessibility review today as I'm interested in this getting landed :)

knyghty commented on django/contrib/admin/options.py:
```diff
@@ -250,7 +250,7 @@ def formfield_for_choice_field(self, db_field, request, **kwargs):
         """
         If the ModelAdmin specifies ordering, the queryset should respect that
-        ordering.  Otherwise don't specify the queryset, let the field decide
+        ordering. Otherwise don't specify the queryset, let the field decide
```
I think we usually try to avoid unrelated cleanup in PRs but I'll let the fellows be the judge, I don't really know.

knyghty commented on django/contrib/admin/options.py:
```diff
@@ -1836,7 +1851,21 @@ def _changeform_view(self, request, object_id, form_url, extra_context):
+
+        actions = self.get_actions(request)
         if request.method == "POST":
+            selected_action = request.POST.get("action", "")
+            if actions and selected_action:
```
I wonder if we could use walrus operator here?

mgaligniana replied:
I think yes! As it is used just for the condition! Thank you!

knyghty commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -149,6 +149,26 @@ That code will give us an admin change list that looks something like this:
 
+.. versionchanged:: 6.0
+
+    The actions will be available also in the detail view but the description should be
+    adjusted to make sense we are talking about only one object::
```
Reads better like this to my native English brain:

"The actions will also be available in the detail view but the description should be adjusted to show we are only talking about a single object."

knyghty commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -149,6 +149,26 @@ That code will give us an admin change list that looks something like this:
+        def make_published(modeladmin, request, queryset):
+            pass
+
+    It will looks like this:
```
```suggestion
    It will look like this:
```

knyghty reviewed (commented):
I had a quick scan of the code but not too deep. I left some comments.

From an accessibility point of view I see no issues here because we're just re-using what already exists.

sarahboyce commented on django/contrib/admin/options.py:
```diff
@@ -1649,7 +1654,17 @@ def response_action(self, request, queryset):
                 return response
             else:
-                return HttpResponseRedirect(request.get_full_path())
+                # If action was executed from change view, redirect to
+                # list view once finished.
+                is_change_view = request.resolver_match.url_name.endswith("change")
+                return HttpResponseRedirect(
+                    reverse(
+                        "admin:%s_%s_changelist"
+                        % (self.opts.app_label, self.opts.model_name)
+                    )
+                    if is_change_view
+                    else request.get_full_path()
+                )
```
So I understand this is a nicer flow for the delete action which is provided by default, however I wonder if this is nicer for the other custom actions :thinking: 
The delete action on the single view is not that useful (as we already have a delete button when you have a delete permission) so I would question if we even should list it

Do you think we should have a boolean like `changelist_only` that can be set on the action? Perhaps defaults to `True` so folks need to opt in to the feature and actions like the delete case might not want to be on a singular level?

mgaligniana replied:
> So I understand this is a nicer flow for the delete action which is provided by default, however I wonder if this is nicer for the other custom actions 🤔

Makes sense. In the last commit I've deleted the logic I had added redirecting to the list view + I've added a success message.

> The delete action on the single view is not that useful (as we already have a delete button when you have a delete permission) so I would question if we even should list it
> Do you think we should have a boolean like `changelist_only` that can be set on the action? Perhaps defaults to `True` so folks need to opt in to the feature and actions like the delete case might not want to be on a singular level?

Brilliant! I've added that `changelist_only` option. I did it a new commit so you can check easily the new changes + new test

sarahboyce reviewed (commented):
Thank you for the updates @mgaligniana 
We need a release note in 6.0 and I have a question about the API design :grin: but I think we're close

mgaligniana commented:
I always forget the release notes! Arggh! 😆

Thank you Sarah! I'll read and apply the comments during this week, thank you!

mgaligniana commented on django/contrib/admin/decorators.py:
```diff
@@ -1,5 +1,10 @@
+    permissions=None,
+    description=None,
+    description_plural=None,
+    changelist_only=True,
```
🎗️ Add new documentation for this new paramenter

mgaligniana commented on django/contrib/admin/options.py:
```diff
@@ -1648,23 +1652,16 @@ def response_action(self, request, queryset):
 
             response = func(self, request, queryset)
 
+            self.message_user(request, "Action executed successfully", messages.SUCCESS)
```
🎗️ For sure it could be better message or more detailed. I'll wait for comments!

edvm commented:
This is so cool, thanks for working on this @mgaligniana , I've been waiting for this feature idk... since 2023? one more time, thank you!!! 🙌 🙌 🙌

sarahboyce reviewed (commented):
We also need a release note
I made a suggestion to update `changelist_only` to `show_on_change_form`, we might want to get other peoples opinions on a range of names

Possible release notes in 6.0 admin section could be:
```
* The new ``show_on_change_form`` argument of
  :func:`~django.contrib.admin.action` allows the action to be available on the
  admin change form view. Defaults to ```False``.

* The new ``description_plural`` argument of
  :func:`~django.contrib.admin.action` specifies a human-readable description
  for actions applied to multiple objects on the admin changelist. Defaults to
  the ``description`` value. This is useful when ``show_on_change_form`` is
  set to ``True``, where ``description`` is then used for a single object on
  the change form.
```

Antoliny0919 commented on docs/ref/contrib/admin/_images/adding-actions-to-the-modeladmin-detail-view.png:
Oops, I wonder why that mention didn’t notify me!
I came across it by chance while looking through some old PRs.
Of course, that’s totally fine!
I’ll add screenshots on the 6th or 7th.

By the way, it's really great work @mgaligniana  :)

mgaligniana commented on docs/.DS_Store:
Ups! Thank you!

sarahboyce commented:
Thank you everyone and @mgaligniana especially for all the work on this PR
I am going to push minor tweaks in prep for a final approval and merge :+1:

sarahboyce commented:
@Antoliny0919 if you have capacity can you regenerate the image with the dropdown? I believe the "Delete" option should not be visible

Antoliny0919 commented:
> @Antoliny0919 if you have capacity can you regenerate the image with the dropdown? I believe the "Delete" option should not be visible

Of course! I'll share it by tomorrow :)

sarahboyce commented:
@Antoliny0919 I might tweak the UI still sorry. I will update you when I'm ready (sorry for the noise)

mgaligniana commented:
If there's anything I can do/help just let me know! Lets go team! 🚀

mgaligniana committed:
Fixed #12090 -- Added admin actions to the admin change form.

Thank you to Benjamin Balder Bach and Jacob Walls for reviews.

Co-authored-by: Sarah Boyce <42296566+sarahboyce@users.noreply.github.com>

sarahboyce commented:
Following the discussion in https://github.com/django/new-features/issues/146, I have pushed the following changes:

- updated the action decorator to use `action_type` rather than a `show_on_change_form` boolean. This now means it is possible to define an action to **only** be on the change form.
- created a new template file for the change form actions. The purpose is to make it easier for folks to customize, e.g. to use buttons in the change form view
- added some deprecations as the method signature changes would break some projects if folks have overridden these methods

@mgaligniana if you can review and test these recent changes, that would be fantastic

mgaligniana commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -467,7 +511,55 @@ The ``action`` decorator
+.. class:: ActionType
+
+  Enum of allowed values for the ``action_type`` parameter of the
+  :func:`~django.contrib.admin.action` decorator .
```
space before ` .`

mgaligniana commented:
I think we have a possible 🐛  here:

At `django/contrib/admin/decorators.py` docstring it says

```
    This is equivalent to setting some attributes (with the original, longer
    names) on the function directly::

        def make_published(self, request, queryset):
            queryset.update(status='p')
        make_published.allowed_permissions = ['publish']
        make_published.short_description = 'Mark selected stories as published'
```

If someone has an action defined with only a short_description (no decorator) it will break

Try running this test:

```python
  def test_undecorated_site_action(self):
      from .admin import site as test_site

      def archive(modeladmin, request, queryset):
          pass

      archive.short_description = "Archive selected items"

      test_site.add_action(archive)
      try:
          self.client.force_login(self.superuser)
          response = self.client.get(
              reverse("admin:admin_views_externalsubscriber_changelist")
          )
          self.assertContains(
              response,
              '<option value="archive">Archive selected items</option>',
              html=True,
          )
      finally:
          test_site.disable_action("archive")
```

The fix could be:

```diff
- if action_type not in func.action_types:
+ if action_type not in getattr(func, "action_types", [ActionType.BULK_ACTION]):
```

sarahboyce commented:
> I think we have a possible 🐛 here:
> [...]

Great find! Added a test and resolved thank you :star:

benjaoming commented on django/contrib/admin/options.py:
```diff
@@ -86,6 +89,11 @@ class ShowFacets(enum.Enum):
 
+class ActionType(enum.Enum):
+    BULK_ACTION = 0
+    SINGLE_ACTION = 1
```
I might have originated these labels, so allow my to scrutinize my own suggestion 😄 I was thinking for a while if it works when adding more options: And I think we should probably name them differently to avoid ambiguity in the future.

The feature proposal of having buttons together with the object tools is likely to go in, so I'll use that as an example... And there might also be other types later... like `CHANGEFORM_BUTTONS_DANGEROUS` or `CHANGELIST_OBJECT_TOOLS`.. I dunno.

Gonna finish quickly instead of a long rant.. but I think the naming would make more sense if it relates to the location instead of an abstraction of its nature:


```suggestion
class ActionType(enum.Enum):
    CHANGELIST_DROPDOWN = 0
    CHANGEFORM_DROPDOWN = 1
```

I know this is less elegant, but then we can add the new feature proposal:

```
class ActionType(enum.Enum):
    CHANGELIST_DROPDOWN = 0
    CHANGEFORM_DROPDOWN = 1
    CHANGEFORM_OBJECT_ACTIONS = 2
```

It's annoying with the changeform/changelist/change naming. Would be nice to call them `LIST_*` and `UPDATE_*` but I think probably better to stick with something familiar ❓ 🤷

benjaoming replied:
I'm tempted to go for something else that's more elegant 💡 I really liked `SINGLE_ACTION` and `BULK_ACTION`. Okay how about:

```
class ActionType(enum.Enum):
    LIST_DROPDOWN = 0
    OBJECT_DROPDOWN = 1
    OBJECT_TOOLS_BUTTON = 2
```

benjaoming replied:
@sarahboyce 

I'm lacking a bit of vocabulary / grasping here... so disclaimer that hopefully I will understand my own input better in a few hours/days:

I think your concern about how to handle forms in `response_action` and `_changeform_view` is 100% valid and to the point of what goes into "action type", i.e. the definition of what we mean here.

We need a way for new action types to be composed in the future so they are handled in the right place.

In that sense, you're right that SINGLE_ACTION and BULK_ACTION seem to be fundamental types. We don't want to add explicit handling of new action types when they're fundamentally doing the same thing in terms of form handling.

benjaoming replied:
So we have this:

* Ability to define widget
* Ability to define location
* Ability to define form handling
* ...+ any other loose aspect that a project or app might add to the interpretation of their own Action Types

I'm becoming more and more tempted to going BACK to the binary enums for this reason. Especially since it would have a very low impact on the current implementation.

benjaoming commented on django/contrib/admin/options.py:
```diff
@@ -1039,7 +1061,28 @@ def _filter_actions_by_permissions(self, request, actions):
         return filtered_actions
 
-    def get_actions(self, request):
+    # RemovedInDjango70Warning: When the deprecation ends, remove.
```
Do we need the deprecation warning?

I know that the interface has changed by adding a new parameter, but it defaults to the same behavior as previously. And it could keep having that default?

benjaoming replied:
(same comment for `get_action_choices`)

sarahboyce replied:
`ModelAdmin.get_actions(request)` is documented with a recommendation that folks can/should override it: https://docs.djangoproject.com/en/6.0/ref/contrib/admin/actions/#django.contrib.admin.ModelAdmin.get_actions
If folks have overridden it, it won't accept the new parameter and their admin will break so we need to add a deprecation. This is the one I feel most strongly about.

`get_action_choices` is not documented but it is not prefixed with `_` to imply it being private. So I chose to deprecate. This probably has a similar use case for being overridden.
I nearly deprecated `response_action` and `get_action` but these were more fiddly and I saw less of a reason to override these so I chose not to deprecate. 

I understand that the decision here is inconsistent. Do you have an opinion here?

benjaoming replied:
Oh! May I should have considered the technical aspect of this before asking! I see clearly this morning what you've crafted is and why it makes sense 👍 I don't have any better way to address my concern _technically_. It seems inevitable that when we change this signature, we have to start having inspections of legacy signatures - which we don't want to have lingering around forever 👍 

My worry was more if we could get rid of the annoyance of fixing a "zero-value" deprecation warning. They most-often bubble up from third-party Django applications.... and stay around for a while 😖

benjaoming replied:
> I understand that the decision here is inconsistent. Do you have an opinion here?

I've never had needs to override `get_action_choices` - I can find some evidence around where people want to change the "empty value" dashes:

* https://stackoverflow.com/questions/35503403/how-to-remove-in-django-admin-action
* https://forum.djangoproject.com/t/modify-action-select-style/27296

but in BOTH cases, good people are VERY clear that it's an internal method since it's undocumented.

Since it's so easy to deprecate along the other method, I think you should stick to your decision here and not pile up more work 👍 

But maybe you know a way forward for refactoring these pseudo-private methods in the longer run, bit by bit? 🧠

benjaoming replied:
Gonna resolve this, just to be clear, there's nothing actionable from my original input!

benjaoming replied:
The "Resolve comment" button doesn't work. But please consider it resolved.

benjaoming commented on django/contrib/admin/templates/admin/change_form.html:
```diff
@@ -37,6 +37,10 @@
+{% if action_form %}
+  {% change_form_admin_actions %}
+  <input type="hidden" name="{{ action_checkbox_name }}" value="{{ object_id|unlocalize }}">
+{% endif %}
```
👍 for keeping the whole interface and for the minimal changes imposed 💚 👏 

We're unconditionally rendering the action_form here - but we could add `{% if is_popup %}` around it as a safeguard? It's because it seems like we're not populating any actions here, anyways:

https://github.com/mgaligniana/django/blob/9f1d70e6cf9a8e04e1468caf97b144b9d110e2bc/django/contrib/admin/options.py#L1085-L1093

sarahboyce replied:
Thank you! Pushed a change and added testing :+1:

benjaoming commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -431,10 +436,49 @@ For example::
+
+Notice that ``description`` and ``description_plural`` were provided. These are
+optional but the admin action will be labeled by ``description`` in the admin
+change form and ``description_plural`` in the admin changelist.
```
```suggestion
change form and ``description_plural`` in the admin change list.
```

nit: why is it "changelist" (one word) and "change form" (two words)? Is this like because of naming consistency?

sarahboyce replied:
It's also inconsistent between `change list` and `changelist`, might be because there is a `ChangeList` class. I will have a chat with the Fellows and try to figure out what I should be saying in these contexts :+1:

sarahboyce replied:
(Updated to `change list` as this is what is already consistently being used in the action docs)

benjaoming commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -431,10 +436,49 @@ For example::
+
+By default, admin actions are available on the changelist page only. You can
+control where an action appears using the ``action_type`` argument of the
+``@admin.action`` decorator.
```
```suggestion
By default, admin actions are available on the changelist page only. You can
now control where an action appears using the ``action_type`` argument of the
``@admin.action`` decorator.
```

(added "now" so it's clear what the change is)

sarahboyce replied:
I can't put now here as it's the normal docs (so it stays the same forever) but "now" can be added to any "versionchanged" note or release note

benjaoming replied:
Misread! I thought I was reading a changelog :face_exhaling:

benjaoming reviewed (commented):
Wow 😍 

I have a worry that a model with a field named `action` would clash with the action select name... which is also named `action`, right?

We could consider a test and if it fails, we could consider prefixing the action form that is submitted ❓

benjaoming commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -478,10 +522,44 @@ The ``action`` decorator
+
+  .. attribute:: SINGLE_ACTION
+
+      The action is available on the admin change form.
```
```suggestion
      The action is available on the admin change form. Saving changes to objects is skipped when an action is run.
```

I think this behavior is important to document. We have a test case for it, thanks for adding that 👍

sarahboyce replied:
Good point! I added `When an action is run, any unsaved changes on the admin change form will be lost.`
But I can change to your original suggestion if preferred :+1:

benjaoming replied:
That's even better :+1:

sarahboyce commented:
> I have a worry that a model with a field named action would clash with the action select name... which is also named action, right?
>
> We could consider a test and if it fails, we could consider prefixing the action form that is submitted ❓

I have tested and added logic for this, thank you :star:

sarahboyce commented:
Hi folks! I pushed up a new commit where I am changing the naming away from `action_type` to be focused on `location`. I think this is my current preference. I would be interested in what @mgaligniana @benjaoming @nessita @jacobtylerwalls think about this and what your preferences are

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1039,7 +1096,27 @@ def _filter_actions_by_permissions(self, request, actions):
+    def _get_actions_with_action_location(
+        self, request, action_location=ActionLocation.CHANGE_LIST
+    ):
+        sig = inspect.signature(self.get_actions)
```
We have to use the safe version of this function from django.utils.inspect in case 3.14 users are using lazy annotations. (and below)

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1039,7 +1096,27 @@ def _filter_actions_by_permissions(self, request, actions):
+                "deprecated. Update the signature to get_actions(self, request, "
+                "action_location=ActionLocation.CHANGE_LIST).",
+                RemovedInDjango70Warning,
+            )
```
django_file_prefixes() here? (and below)

jacobtylerwalls commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -362,27 +364,35 @@ including any :ref:`site-wide actions <adminsite-actions>`.
+        The keyword arguments ``location`` was added. The return type was
+        changed to a dictionary where keys the values are :class:`Action`
+        objects, previously these were ``(function, name, description)``
+        tuples.
```
arguments -> argument
and "keys the values" -> maybe, "a dictionary, as described earlier."?

jacobtylerwalls commented on django/contrib/admin/decorators.py:
```diff
@@ -23,6 +35,11 @@ def decorator(func):
+            func.plural_description = description_plural
+        elif description is not None:
+            func.plural_description = description
+        func.locations = location if isinstance(location, Iterable) else [location]
```
I always make exceptions for strings when using `Iterable`, did you consider that here? (Someday we might change the enum to a StrEnum.)

sarahboyce replied:
I can make it so it must always be an iterable. I think @benjaoming suggested allowing both types to be accepted

benjaoming replied:
@sarahboyce I think @jacobtylerwalls means this:

```
>>> isinstance("test", Iterable)
True
```

It seems to me like it's a Django convention to allow both a single string and an iterable, thinking especially about `order_by("date")` and `order_by(["date", "name"])`

The reasoning here would be: Many/most needs would be covered by a single location (and thus it's supported as a single string), but the flexibility provided by supporting multiple locations is going to provide us some big wins. I don't wanna have opinions about whether this is a good or bad pattern, I was more trying to lean into that existing convention.

jacobtylerwalls replied:
I assumed strings didn't work, so I wanted the failure to happen at the same time as the isinstance check. If they work, then great, no change needed. But I don't see a test -- can we add one for strings? (Was there one before, and this comment made you delete it? Sorry!)

sarahboyce replied:
Thank you, I have done an update.
We don't have type checking for any of the other attributes so I am not keen on adding handling for failure modes here

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1888,11 +2039,39 @@ def _changeform_view(self, request, object_id, form_url, extra_context):
         if request.method == "POST":
+            if action_form and request.POST.get(action_form["action"].html_name, ""):
+                queryset = self.model._default_manager.get_queryset()
+                if response := self.response_action(
```
From the Claude review above:
> select_across security concern on change form

I think we have to consider forged POSTs: we should definitely either raise or force `select_across=False` and add a test.

jacobtylerwalls commented on django/contrib/admin/options.py:
```diff
@@ -1888,11 +2039,39 @@ def _changeform_view(self, request, object_id, form_url, extra_context):
+        actions = self._get_actions_with_action_location(
+            request, action_location=ActionLocation.CHANGE_FORM
+        )
+        if actions and not add:
```
On the change list, when I click "Run" while "Select an option" is the value, I get a message: "Items must be selected in order to perform actions on them. No items have been changed."

I would expect a message (different wording) when the same is attempted on the single object form. Instead, I get "The `<object>` was changed successfully.'

jacobtylerwalls reviewed (commented):
I like how location represents what the framework actually does (renders an action as a choice in a location). Whether it's bulk-friendly or not is left up to the developer. 👍

benjaoming commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -362,27 +364,35 @@ including any :ref:`site-wide actions <adminsite-actions>`.
-    and the values are ``(function, name, short_description)`` tuples.
+    This returns a dictionary of actions allowed for the specific
+    ``action_location``. The keys are action names, and the values are
+    :class:`Action` objects.
```
Thanks for taking the time to think of a new idea and try it out. This seems like a worthwhile improvement to a long-term API 🤞

benjaoming commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -362,27 +364,35 @@ including any :ref:`site-wide actions <adminsite-actions>`.
 
+    .. versionchanged:: 6.1
+
+        The keyword argument ``location`` was added. The return type was
```
Wait isn't this `action_location`? 😵‍💫

benjaoming reviewed (approved):
+1 on the `location` change 👏 (which addressed what was my primary focus)

I'll see if I have time to re-visit, but please count me as a happy reviewer 🙏

jacobtylerwalls commented on docs/releases/6.1.txt:
```diff
@@ -637,6 +611,16 @@ Miscellaneous
   ``"sha1"`` to ``"sha256"`` in Django 7.0. Pass an explicit ``algorithm``
   to silence the deprecation warning.
 
+* Overriding ``ModelAdmin.get_actions()`` without the new ``action_location``
+  parameter is deprecated.
+
+* Unpacking or indexing the dictionary values of the
+  ``ModelAdmin.get_actions()`` return value is deprecated. Use
+  :class:`~django.contrib.admin.Action` attributes instead.
+
+* Overriding ``ModelAdmin.get_action_choices()`` without the new
+  ``action_location`` parameter is deprecated.
+
```
Need corresponding entries in docs/internals/deprecation.txt.

jacobtylerwalls commented on django/contrib/admin/templates/admin/change_form.html:
```diff
@@ -37,6 +37,10 @@
 {% block content %}<div id="content-main">
 <form {% if has_file_field %}enctype="multipart/form-data" {% endif %}{% if form_url %}action="{{ form_url }}" {% endif %}method="post" id="{{ opts.model_name }}_form" novalidate>{% csrf_token %}{% block form_top %}{% endblock %}
 <div>
+{% if action_form and not is_popup %}
```
Do we want to respect `actions_on_bottom = True`?

sarahboyce replied:
I think we can add this as a new feature if wanted. I have clarified the docs that this is only for the admin change list page
